ppx_view/lib/expand.ml

```diff
@@ -0,0 +1,337 @@
+       (* why is this needed? *)
+       let expr, vars = translate_tuple ~loc:ploc patts in
+       ( apply [expr]
+       , vars )
+     | _ ->
+       let expr, vars = translate_pattern ~err_loc:loc patt in
+       ( apply [expr]
+       , vars ))
+  | Some (Ppat_or (first, second)) ->
+    let expr_first, vars_first = translate_pattern ~err_loc:loc first in
+    let expr_second, vars_second = translate_pattern ~err_loc:loc second in
+    if same_variables ~err_loc:loc vars_first vars_second then
+      let f = Builder.view_lib_choice ~loc in
+      ( Builder.Exp.apply_lident ~loc f [expr_first; expr_second]
+      , vars_first )
+    else
+      Error.or_pattern_variables_differ ~loc
+  | Some (Ppat_record (fields, _closed)) ->
+    let exprs_vars = List.map ~f:(translate_record_field ~loc) fields in
+    let exprs, vars = List.split exprs_vars in
+    ( Builder.Exp.view_lib_sequence ~loc exprs
+    , List.flatten vars )
+  | Some (Ppat_array patts) ->
+    translate_array ~loc patts
+  | Some (Ppat_constraint (patt, ctyp)) ->
+    let expr, vars = translate_pattern ~err_loc:loc patt in
+    ( Builder.expression ~loc (Expression_desc.pexp_constraint expr ctyp)
+    , vars )
+  | Some (Ppat_open (lid, patt)) ->
+    let expr, vars = translate_pattern ~err_loc:loc patt in
+    ( Builder.expression ~loc
+        (Expression_desc.pexp_open Override_flag.override lid expr)
+    , vars )
+  | Some (Ppat_variant _) ->
+    Error.unsupported_pattern ~loc "polymorphic variants"
+  | Some (Ppat_type _) ->
+    Error.unsupported_pattern ~loc "sub types"
+  | Some (Ppat_lazy _) ->
+    Error.unsupported_pattern ~loc "lazy values"
+  | Some (Ppat_unpack _) ->
+    Error.unsupported_pattern ~loc "first class modules"
+  | Some (Ppat_exception _) ->
+    Error.unsupported_pattern ~loc "exceptions"
+  | Some (Ppat_extension _) ->
+    Error.unsupported_pattern ~loc "extension points"
+
+and translate_patterns ~err_loc patts =
+  let exprs, vars =
+    List.split (List.map ~f:(translate_pattern ~err_loc) patts)
+  in
+  (exprs, List.flatten vars)
+
+and translate_tuple ~loc patts =
+  let exprs, vars = translate_patterns ~err_loc:loc patts in
+  match List.length exprs with
+  | (2 | 3 | 4) as len ->
+    let f =  Builder.view_lib_ident ~loc (Printf.sprintf "tuple%d" len) in
+    ( Builder.Exp.apply_lident ~loc f exprs
+    , vars )
+  | len ->
+    assert (len > 4);
+    let f = inline_tuple_function ~loc len in
+    ( Builder.Exp.apply ~loc f exprs
+    , vars )
+
+and translate_record_field ~loc (label, patt) =
+  let (label_loc, label) = Deconstructor.longident_loc ~loc label in
+  match label with
+  | Lident label ->
+    let expr, vars = translate_pattern ~err_loc:label_loc patt in
+    let f = Builder.lident_loc ~loc:label_loc (label ^ "'match") in
+    ( Builder.Exp.apply_lident ~loc:label_loc f [expr]
+    , vars )
+  | _ ->
+    (* Couldn't that be seen as an open and the field
+       translated to [Some_module.x'match]? *)
```

Comment:
That seems good to me, but we should probably use the same module name for all fields. i.e. we could use the same rule as OCaml: for `{A.x;y}`, we'd use module path `A` for all fields.

Reply:
Yeah that's what I had in mind indeed, otherwise it would be too far from the semantic of the regular qualified field name. I'll add an issue about this and will try to fix it later. I think it's more important to get it working for our internal use first and I still have to port parseview!

Reply:
Sound good!